### ISSUE TITLE
Does not match S3 semantics on multipart POST uploads

### ISSUE THREAD

**Issue by phihag** [minio/minio#6443]
We are uploading files with a command like

```
curl -X POST -H 'content-type: multipart/form-data; boundary=Boundary-45280d2a-f82c-48fe-b042-79b0b17406b9' -H Expect: -d '--Boundary-45280d2a-f82c-48fe-b042-79b0b17406b9
Content-Disposition: form-data; name="x-amz-algorithm"

AWS4-HMAC-SHA256
--Boundary-45280d2a-f82c-48fe-b042-79b0b17406b9
Content-Disposition: form-data; name="x-amz-credential"

OURLOCALKEY/DEVSETUP/20180910/local/s3/aws4_request
--Boundary-45280d2a-f82c-48fe-b042-79b0b17406b9
Content-Disposition: form-data; name="x-amz-date"

20180910T125644Z
--Boundary-45280d2a-f82c-48fe-b042-79b0b17406b9
Content-Disposition: form-data; name="x-amz-signature"

2a031447cf89a94249f74e7d3a24cc862761425964bbd24fb8b7ee7cab6aab66
--Boundary-45280d2a-f82c-48fe-b042-79b0b17406b9
Content-Disposition: form-data; name="policy"

eyJleHBpcmF0aW9uIjogIjIwMTgtMDktMTBUMTM6NTY6NDRaIiwgImNvbmRpdGlvbnMiOiBbeyJidWNrZXQiOiAiYnhuLXRvbmllY2xvdWQtZGV2LXVwbG9hZCJ9LCB7ImtleSI6ICIzNjc5OWU1Yi0xMjFhLTQ2ZDUtYTI5YS1kZGY3NmUwNWFjOTMifSwgeyJ4LWFtei1hbGdvcml0aG0iOiAiQVdTNC1ITUFDLVNIQTI1NiJ9LCB7IngtYW16LWNyZWRlbnRpYWwiOiAiQk9YSU5FTE9DQUwvREVWU0VUVVAvMjAxODA5MTAvbG9jYWwvczMvYXdzNF9yZXF1ZXN0In0sIHsieC1hbXotZGF0ZSI6ICIyMDE4MDkxMFQxMjU2NDRaIn1dfQ==
--Boundary-45280d2a-f82c-48fe-b042-79b0b17406b9
Content-Disposition: form-data; name="key"

36799e5b-121a-46d5-a29a-ddf76e05ac93
--Boundary-45280d2a-f82c-48fe-b042-79b0b17406b9
Content-Disposition: form-data; name="file"; filename="36799e5b-121a-46d5-a29a-ddf76e05ac93"
Content-Type: application/octet-stream

this is a file

--Boundary-45280d2a-f82c-48fe-b042-79b0b17406b9--
' http://localhost:9000/ourbucket
```

This works fine on S3 and various other S3 mocks (for instance localstack or fake-s3) and returns a 204 (or 200), but fails with minio with HTTP 400 and the output:

```
<?xml version="1.0" encoding="UTF-8"?>
<Error><Code>MissingFields</Code><Message>Missing fields in request.</Message><Resource>/ourbucket</Resource><RequestId>15530B8856316D89</RequestId><HostId>3L137</HostId></Error>
```

If minio aspires to be compatible to S3, then it would be nice to support multipart post requests as well.

It would also be nice if minio could be more verbose in the error message and mention _which_ field is missing.

I'm using the minio docker image tagged latest, which seems to be RELEASE.2018-09-01T00-38-25Z
 , on debian stretch amd64.

**Comment by aead** [comment]
@phihag The server rejects the request because it treats the `x-am-credential` string as invalid.
Can you confirm that your AccessKeyIDs have the form `x/y` like `OURLOCALKEY/DEVSETUP` ?!

**Comment by phihag** [comment]
Thank you very much! Replacing the `/` with `-` fixed our problem. (We used another S3 mock before, and that used slashes in the access key ID).

**Comment by aead** [comment]
AFAICS `/` seems to be a valid character for S3 access keys. (At least AWS accepts such keys making it de-facto "expected behavior").
Reopening this to track the issue (addressed by #6451)

**Comment by Arkemlar** [comment]
I found this issue by googling for error "Missing fields in request". In my case problem was that when I used `openssl rand -base64 30` to generate secret key it randomly generated key with slash `/` at the end. Thanks to discussion above I realised that the problem is slash at the key's end. Removed this slash in key and everything worked fine.
